### PR TITLE
🐣🐇🐣 new package: kind-of 🎅

### PR DESCRIPTION
**Pull request by claasahl**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

If removing a declaration:
- [ ] If a package was never on DefinitelyTyped, you don't need to do anything. (If you wrote a package and provided types, you don't need to register it with us.)
- [ ] Delete the package's directory.
- [ ] Add it to `notNeededPackages.json`.

### CODE REVIEW

**Comment by colbywhite** [comment]
With 31.8M weekly downloads, `kind-of` certainly seems like a popular package. 

I'm curious however if it is relevant in _typescript_ projects. It seems like the package is all about doing type checks, which is what typescript does already. in a typescript project, what's the use case for needing a package like `kind-of`? Not saying this shouldn't be merged, but I'm just curious the use case for it.

**Comment by claasahl** [comment]
Thanks for your comment, @colbywhite

I agree that `kind-of` is not likely to be useful in TS projects. Even more so, when one starts a TS-project from scratch. However, given the popularity of `kind-of`, it is likely to show up in existing code bases that will hopefully soon be transitioned to TS 😉 

Despite my guess that `kind-of` (incl. these types) is not likely to remain in TS-projects for long, my hope is that there are enough people / projects out there which can benefit from these types.

**Review by PranavSenthilnathan — CHANGES_REQUESTED** [review]

File: types/kind-of/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+declare function kind_of(thing: Int8Array): "int8array";
+declare function kind_of(thing: Uint8Array): "uint8array";
+declare function kind_of(thing: Uint8ClampedArray): "uint8clampedarray";
+declare function kind_of(thing: Int16Array): "int16array";
+declare function kind_of(thing: Uint16Array): "uint16array";
+declare function kind_of(thing: Int32Array): "int32array";
+declare function kind_of(thing: Uint32Array): "uint32array";
+declare function kind_of(thing: Float32Array): "float32array";
+declare function kind_of(thing: Float64Array): "float64array";
+declare function kind_of(thing: any): "object" | string;
```
[the remark this answers is not in the document]
Comment (claasahl) [reply types/kind-of/index.d.ts]: Thanks for the pointer, @PranavSenthilnathan 

`"object"` was originally intended as a hint (i.e. this thing that you are testing is _probably_ an "object"). `string` was meant to "catch" those cases where my overloads have diverged from the actual outputs from `kind-of` (e.g. due to incorrect mapping on my behalf or possible updates to the package `kind-of`). To make this intend clearer, I have split this line into two.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e39d217dde6e29507f7ab55710a08b445e68b54d/types/kind-of/index.d.ts#L48-L49

I am aware that it is not recommended to use `Object`. Given that `kind-of` will recognize `Object.prototype` as `"object"`, `Number.prototype` as `"number"`, `String.prototype` as `"string"` (see [RunKit + npm: kind-of](https://runkit.com/embed/8dccvqtplzqg)), how should one resolve this? Should one simply drop references to `Object` and such, even though they actually seem correct? Perhaps, I am not seeing something here.

[the remark this answers is not in the document]
Comment (claasahl) [reply types/kind-of/index.d.ts]: ... just pushed a work-around.

**Review by PranavSenthilnathan — APPROVED** [review]

**Pull request merged**